File: frontend/src/app/post/services/post.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Post, PostRequest, Tag } from '../models/post.model';

const API_URL = environment.postApiUrl;

export interface PostFilters {
  tagName?: string;
  titleSearch?: string;
  authorId?: number;
  myPosts?: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class PostService {
  constructor(private http: HttpClient) { }

  getAllPosts(page: number = 0, size: number = 10, filters?: PostFilters): Observable<any> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());

    if (filters) {
      if (filters.tagName) {
        params = params.append('tagName', filters.tagName);
      }
      if (filters.titleSearch) {
        params = params.append('titleSearch', filters.titleSearch);
      }
      if (filters.authorId !== null && filters.authorId !== undefined) {
        params = params.append('authorId', filters.authorId.toString());
      }
      if (filters.myPosts !== null && filters.myPosts !== undefined) {
        params = params.append('myPosts', filters.myPosts.toString());
      }
    }

    return this.http.get(`${API_URL}/posts`, { params });
  }

  getPostById(id: number): Observable<Post> {
    return this.http.get<Post>(`${API_URL}/posts/${id}`);
  }

  createPost(postRequest: PostRequest): Observable<Post> {
    return this.http.post<Post>(`${API_URL}/posts`, postRequest);
  }

  updatePost(id: number, postRequest: PostRequest): Observable<Post> {
    return this.http.put<Post>(`${API_URL}/posts/${id}`, postRequest);
  }

  deletePost(id: number): Observable<any> {
    return this.http.delete(`${API_URL}/posts/${id}`);
  }

  getPostsByAuthor(authorId: number): Observable<Post[]> {
    return this.http.get<Post[]>(`${API_URL}/posts/author/${authorId}`);
  }

  searchPosts(query: string, page: number = 0, size: number = 10): Observable<any> {
    return this.http.get(`${API_URL}/posts/search?query=${query}&page=${page}&size=${size}`);
  }

  getPostsByTag(tagName: string, page: number = 0, size: number = 10): Observable<any> {
    return this.http.get(`${API_URL}/posts/tag/${tagName}?page=${page}&size=${size}`);
  }

  updatePostStatus(id: number, status: string): Observable<Post> {
    return this.http.put<Post>(`${API_URL}/posts/${id}/status?status=${status}`, {});
  }

  votePost(id: number, upvote: boolean): Observable<Post> {
    return this.http.post<Post>(`${API_URL}/posts/${id}/vote?upvote=${upvote}`, {});
  }

  getAllTags(): Observable<Tag[]> {
    return this.http.get<Tag[]>(`${API_URL}/tags`);
  }

  createTag(name: string): Observable<Tag> {
    return this.http.post<Tag>(`${API_URL}/tags?name=${name}`, {});
  }
} 

File: backend/post-service/src/main/java/com/facebook/post/service/PostService.java
package com.facebook.post.service;

import com.facebook.post.model.PostStatus;
import com.facebook.post.payload.PostDTO;
import com.facebook.post.payload.request.PostRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface PostService {
    Page<PostDTO> getAllPosts(Pageable pageable);
    
    Page<PostDTO> getAllPostsFiltered(Pageable pageable, String tagName, String titleSearch, Long authorId);
    
    PostDTO getPostById(Long id);
    
    PostDTO createPost(Long authorId, PostRequest postRequest);
    
    PostDTO updatePost(Long id, Long userId, PostRequest postRequest);
    
    void deletePost(Long id, Long userId);
    
    List<PostDTO> getPostsByAuthor(Long authorId);
    
    Page<PostDTO> searchPostsByTitle(String query, Pageable pageable);
    
    Page<PostDTO> getPostsByTag(String tagName, Pageable pageable);
    
    PostDTO updatePostStatus(Long id, Long userId, PostStatus status);
    
    PostDTO votePost(Long id, Long userId, boolean upvote);
    
    void updatePostVoteCount(Long postId, int voteCount);
} 